### PR TITLE
DB-11718 Refine index lookup cost estimation and evaluate constant string

### PR DESCRIPTION
**Pull request by ascend1**
## Short Description
This PR includes two changes:
* evaluate constant string to string conversion in binding
* index lookup cost tuning

## Long Description
In investigating query hash `SMopSK5W`, three problems are identified (for detailed explanation, see Jira). This PR addresses the first two issues.

1. So far, we don't evaluate a cast node if it's from string to string, for example, `cast('abc' as char(4))`. Predicate selectivity estimation is affected if a predicate contains such a cast node as one of its operand. Since it's not a constant node, we will go for default estimation rather than ask statistics for a concrete value.

The reason we don't evaluate this kind of cast might be that we could cast a constant string to a narrower type, leading to data truncation. The truncation itself is totally fine, but according to standard, we should raise a warning in this case and send this warning together with the final result set to client. This is something not really possible in binding phase. However, early evaluation should be possible for casts whose target types are wider.

2. Currently, our index lookup cost has a penalty of 4000 per row. There are two issues about this. Firstly, the penalty is too high compared to actual execution times. Secondly, it's an over simplification because index lookup happens in a batched manner and could have some parallelism. Batch size and the maximum number of batches that can run concurrently are controlled by settings in `SQLConfiguration`. This change tunes the index lookup cost to consider these parameters and lower lookup penalty.

## How to test
Note that this PR currently doesn't give a better plan for `SMopSK5W`. Problem 3 described in Jira is preventing the tuned plan to be chosen.

For change 1, casting a string literal to a not-narrower type should be evaluated right in binding phase. However, this difference is not observable on client side. See `StringTypesToStringTypesConversionIT` for more details about this.

For change 2, we should see IndexLookup plans be chosen even when the number of lookup rows are 5 digits. Based on experiments on cloud, looking up 65K rows are still fast, which is just 1.2x - 1.3x times slower than scanning but not looking up the same number of rows. See adaptions in `testDefaultSelectivityFactorHint` for this part. Now looking up 27K rows should be considered faster than doing full table scan in its original setup.

### CODE REVIEW

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2664/

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2681/

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2684/

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 1]
```diff
@@ -67,7 +67,8 @@
 public class ScanCostFunction{
     public static final Logger LOG = Logger.getLogger(ScanCostFunction.class);
 
-    private static final int OLTP_INDEXLOOKUP_LATENCY_FACTOR = 10;
+    private static final int OLTP_INDEXLOOKUP_COST_PER_ROW = 40;
+    private static final int OLAP_INDEXLOOKUP_COST_PER_ROW = 80;
```
Comment (msirek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:71]: I'm not sure we want to used fixed per-row costs.  Shouldn't this be like local and remote costs which are scaled based on the row width?  For example: totalRowCount*totalSelectivity*remoteLatency*(1+**colSizeFactor**/1024d);

**Review by martinrupp — APPROVED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java (hunk 1 above)
Comment (dgomezferro) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:71]: Can we at least open a JIRA to tackle this at a later point?

**Review by dgomezferro — APPROVED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 2]
```diff
@@ -541,10 +559,12 @@ public void generateCost(long numFirstIndexColumnProbes) throws StandardExceptio
...
             scanCost.setIndexLookupCost(-1.0d);
         } else {
-            double lookupCostPerRow = isOlap ? (openLatency+closeLatency) : ((openLatency+closeLatency) / OLTP_INDEXLOOKUP_LATENCY_FACTOR);
-            lookupCost = totalRowCount*filterBaseTableSelectivity*lookupCostPerRow;
-            scanCost.setIndexLookupRows(Math.round(filterBaseTableSelectivity*totalRowCount));
-            scanCost.setIndexLookupCost(lookupCost+baseCost);
+            double numLookupRows = totalRowCount * filterBaseTableSelectivity;
+            double numLookupBatches = Math.max(numLookupRows / indexBatchSize, 1);
+            double batchCost = Math.min(numLookupRows, indexBatchSize) * getIndexLookupCostPerRow() + openLatency + closeLatency;
+            lookupCost = batchCost * Math.max(numLookupBatches / indexLookupBlocks, 1);
```
Comment (hatyo) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:565]: I would extract `Math.max(numLookupBatches / indexLookupBlocks, 1)` into a separate variable:

```
double serialBatchesCount = Math.max(numLookupBatches / indexLookupBlocks, 1);
lookupCost = batchCost * serialBatchesCount;
```

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java (hunk 1 above)
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:71]: Yes, I'll make new commits to improve here. Jira: https://splicemachine.atlassian.net/browse/DB-11795

**Review by hatyo — CHANGES_REQUESTED** [review]
Very minor comments about documentation / refactoring from variable names for readability.

**Comment by ascend1** [comment]
This PR is split into two different PRs: DB-11795 and DB-11829.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 3]
```diff
@@ -98,6 +99,14 @@
     // positions of columns used in estimating selectivity but missing real statistics
     private final HashSet<Integer> usedNoStatsColumnIds;
 
+    // when index lookup is needed, the maximum number of rows that can be fired as a batch
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:102]: Done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 4]
```diff
@@ -98,6 +99,14 @@
...
+    // controled by splice.index.batchSize, default value in SQLConfiguration
+    private final int indexBatchSize;
+
+    // when index lookup is needed, the maximum number of batches that can be fired concurrently
+    // controled by splice.index.numConcurrentLookups, default value in SQLConfiguration
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:107]: Done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 5]
```diff
@@ -98,6 +99,14 @@
...
+    private final int indexLookupBlocks;
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:108]: Done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 6]
```diff
@@ -137,7 +146,12 @@
      * @param scanCost
      * @param resultColumns
      * @param scanRowTemplate
+     * @param baseColumnsInScan
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:149]: Done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java [hunk 7]
```diff
@@ -541,10 +559,12 @@ public void generateCost(long numFirstIndexColumnProbes) throws StandardExceptio
             scanCost.setIndexLookupRows(-1.0d);
             scanCost.setIndexLookupCost(-1.0d);
         } else {
-            double lookupCostPerRow = isOlap ? (openLatency+closeLatency) : ((openLatency+closeLatency) / OLTP_INDEXLOOKUP_LATENCY_FACTOR);
-            lookupCost = totalRowCount*filterBaseTableSelectivity*lookupCostPerRow;
-            scanCost.setIndexLookupRows(Math.round(filterBaseTableSelectivity*totalRowCount));
-            scanCost.setIndexLookupCost(lookupCost+baseCost);
+            double numLookupRows = totalRowCount * filterBaseTableSelectivity;
+            double numLookupBatches = Math.max(numLookupRows / indexBatchSize, 1);
```
[the remark this answers is not in the document]
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:563]: Done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java (hunk 2 above)
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ScanCostFunction.java:565]: Done.

**Pull request closed without merging**